extensions/vault/runtime/src/main/java/io/quarkus/vault/VaultService.java

```diff
@@ -0,0 +1,13 @@
+package io.quarkus.vault;
+
+import java.util.Map;
+
+import io.quarkus.vault.service.dto.VaultDynamicDbCredentials;
+
+public interface VaultService {
+
+    Map<String, String> readSecret(String path);
```

Comment:
@vsevel I'd like to discuss if this how we want the users to read a secret.
Can you please explain a bit what is `path` here, some kind of `namespace` in the vault ? And what is a map key (I guess the value is a secret :-) ) ? 
One thing I doubt is a good idea from the security point of view is fetching all the secrets for a given `path` when a user may only want a single one.

Reply:
yes a path is a namespace. at a given path you have data, which is map of string/string
this is consistent with https://www.vaultproject.io/api/secret/kv/kv-v1.html#read-secret
you do not get a key, but all keys at a given path. I do not know how you would get just one key. see also https://www.vaultproject.io/api/secret/kv/kv-v1.html#create-update-secret

> One thing I doubt is a good idea from the security point of view is fetching all the secrets for a given path when a user may only want a single one.

well the only thing you could do is make the VaultService api more restrictive, with a String readSecret(String path, String key). but then you would have to provide the bigger readSecret(String path) anyway. so from a security standpoint you would not gain anything.

security comes from how you organize your vault. you would typically secrets that are closely related into a given path. then have different paths, all with different access policies. then on top of that you create roles aggregating several policies (one or more for each application).

if a user can fetch secrets he is not supposed to see, that is a problem with how you organize your vault, not with the VaultService api.

Reply:
@vsevel, I'm not sure this conversation is resolved. The read secret operation example shows a single secret being returned. Can you please type an example showing a more complete response ?
The other thing, this `path`, should it be a configuration property ? How else the user code will know the path value ? Hard coding in the code is not ideal.
Should `key` be configurable ?
Even if a map is returned, it does not mean the user should not have one or two simpler shortcut options. It is normal for interface APIs to have it.

Reply:
now that I resolved the native issues, I can focus again on the ongoing discussions.

> just keep the current Map<String, String> readSecret, but: Please introduce a 'dto' object wrapping the map, say, VaultSecrets

actually VaultService does not take or return any dto. it has only 2 public methods:
```
    Map<String, String> readSecret(String path);
    Properties getDynamicDbCredentials(String dbrole);
```

I have moved the VaultService interface to a special io.quarkus.vault.api package, to make it clear what is public, and what is internal.
in particular, remember that VaultClient is an internal implementation detail. design decisions that drive VaultClient do not necessarily apply to VaultService.
VaultService is used by app developers. VaultClient is used internally by the vault extension to call the hashicorp vault process.
I also moved the client stuff to a new io.quarkus.vault.client package. does it make sense? I can also call it io.quarkus.vault.internal.client, to reinforce this. but that sounds overkill.

> Next, should the key be configurable ? One option is, the key is calculated from the request context, that is fine, but if it is a static value then IMHO we need a configuration property.

people would typically store many keys at a particular path, not just one. it would be bizarre to say that there was a special key.

Reply:
OK. lets finalize it. Few comments:
- please at least rename `readSecret` to `readSecrets` given that all of them are returned and to keep an option for a `readSecret` shortcut returning a single value open
- I'd vote for an 'internal' package qualifier for `VaultClient`, otherwise, at least myself, I'm finding it confusing, you know I was hoping to `VaultService` be implemented in terms of `VaultClient` with all the code around dealing with the `VaultService` only, but if you see it all differently then indeed lets at least make it clear that `VaultClient` is for the internal work only :-)
- Absolutely, there should be a single type of the http client only (let it be okhttp for now), why there would be more than one...And please consider simply having `VaultServiceImpl` - so as not to keep renaming when the 'ohhttp' will get replaces with jax-rs or apache http client :-)
- I agree that the right way forward is indeed to have `VaultKvService` and `VaultDbService` to support the other stores in the future and avoid mixing everything in the same interface. But as I said earlier, I'm not sure about a dedicated `VaultAuthService` - would users use it on its own ? No, as we've agreed earlier the login is implicit, and the logout can be done in the Closeable close(). And to keep it simple, please consider just having the `VaultServiceImpl`, implementing both `VaultKvService` and `VaultDbService`.   
Thanks

Reply:
> please at least rename readSecret to readSecrets given that all of them are returned and to keep an option for a readSecret shortcut returning a single value open

done

> I'd vote for an 'internal' package qualifier for VaultClient - but if you prefer to keep where it is now then it is fine.

done

> Absolutely, there should be a single type of the http client only (let it be okhttp for now), why there would be more than one...

done

> I agree that the right way forward is indeed to have VaultKvService and VaultDbService to support the other stores in the future and avoid mixing everything in the same interface.

done

> But as I said earlier, I'm not sure about a dedicated VaultAuthService - would users use it on its own ?

this would serve 2 purposes:
- I need it as a singleton because it is keeping the current token with information to renew it if necessary
- I am still convinced that there is a use case for direct vault access (ie the VaultClient operations). and if we open that up, people will need a client token, which would be provided by the VaultAuthService.
I have always believed that there was room for 2 abstraction levels: service level with the magic, and client level with direct access to vault, such as https://github.com/BetterCloud/vault-java-driver
people would use one or the other level, or both in the same project if some operations are not provided at the service level. they would always have the option of going to the client level.
right now, if everything they need is not in the quarkus vault service level, they would have to bring vault-java-driver in their project for the extra operations they are missing. and they would have to do auth themselves. everything the extension does very well.

but we have had this discussion several times. I can see you have not been convinced so far. and I have refactored the code so many times. so let's move on.

> And to keep it simple, please consider just having the VaultServiceImpl, implementing both VaultKvService and VaultDbService

conceptually I need the vault client and the vault auth manager to be singletons. since they are cdi objects, I decided to move all business logic into a set of managers, held by a single VaultManager. now VaultKvService and VaultDbService are just empty shells delegating to managers. there is enough logic to separate code into different kv manager from db.

> perhaps all of these service interfaces should live in the vault.services package, instead of vault.api

I have reorganized packages:
- io.quarkus.vault: cdi objects
- io.quarkus.vault.internal
- io.quarkus.vault.config

hopefully this should be clearer now.

> the configuration property for the VaultKvService key to be used to retrieve an individual secret from a map; and I guess the same for the dbrole for the VaultDbService

secrets will be fetched using the vault config source: configure a quarkus.vault.secret-config-path and everyting at that path will be mounted as properties in MP Config.
for all the other uses cases, the developer will user VaultKvService by providing the path. I do not think we should do an extra property. this would be confusing with quarkus.vault.secret-config-path
the dbrole is a property of the datasource. see VaultDataSourceRuntimeConfig

how does it sound?

Reply:
@vsevel - I meant a `secret key` property for getting a secret from the map